### PR TITLE
feat/clean unreferenced files startup

### PR DESCRIPTION
**Pull request by pedrotambo**
## Description

This PR is for adding the ability to delete unreferenced files in the content storage. That is, all the files that are not able to be index by the database, are deleted.
This functionality is needed to clean up Catalyst's disks, which have a lot of old heavy files like snapshots.
There was added a mechanism to run the Catalyst in `maintenance` mode. If you run the Catalyst in this mode, the Catalyst will:
- Will start only needed components (small subset of the normal ones)
- Run DB migrations.
- Delete files that are unreferenced in database to free up space.
- Migrate content folder structure to the new structure (functionality added few months ago).
- After that, the Catalyst will shutdown. Then you can start it in normal mode.

Why is this run in maintenance mode? 
Because we don't want it to be run in each startup. It may add a lot of time to bootstrap time and make use of a lot of disk I/O usage and lead to the degradation of it.

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change). If there's an API change, link the pull request in the [API spec repository](https://github.com/decentraland/catalyst-api-specs) and the accepted [DAO governance poll](https://governance.decentraland.org/)
- [X] I have read the [**CONTRIBUTING**](https://github.com/decentraland/catalyst/blob/main/docs/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: content/src/types.ts [hunk 1]
```diff
@@ -80,6 +80,16 @@ export type AppComponents = {
   repository: Repository
 }
 
+export type MaintenanceComponents = {
```
Comment (agusaldasoro) [inline content/src/types.ts:83]: love this!

File: content/test/integration/logic/delete-unreferenced-files.spec.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+// import path from 'path'
+// import { EnvironmentConfig } from '../../../src/Environment'
+// import { deleteUnreferencedFiles } from '../../../src/logic/delete-unreferenced-files'
+// import { bufferToStream } from '../../../src/ports/contentStorage/contentStorage'
+// import { ServiceImpl } from '../../../src/service/ServiceImpl'
+// import { makeNoopValidator } from '../../helpers/service/validations/NoOpValidator'
+// import { loadStandaloneTestEnvironment, testCaseWithComponents } from '../E2ETestEnvironment'
+// import { buildDeployData } from '../E2ETestUtils'
+// import { getIntegrationResourcePathFor } from '../resources/get-resource-path'
+it('i am a test', () => {})
```
Comment (agusaldasoro) [inline content/test/integration/logic/delete-unreferenced-files.spec.ts:13]: 😂

File: content/src/logic/delete-unreferenced-files.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  for await (const storageFileId of storageFileIds) {
+    if (!fileHashes.has(storageFileId)) {
+      await queue.add(async () => {
+        logger.debug(`Deleting: ${storageFileId}`)
+        // await components.storage.delete([storageFileId])
+      })
+      numberOfDeletedFiles++
+    }
+  }
+  // To Do: Log size released?
```
[the remark this answers is not in the document]
Comment (menduz) [reply content/src/logic/delete-unreferenced-files.ts:70]: Metrics are not allowed here bc there may not be any server running. It must print the result to stderr

File: content/src/entrypoints/run-local-database.ts [hunk 4]
```diff
@@ -13,8 +14,32 @@ main()
   .then(() => console.log('Done!'))
   .catch((error) => console.error(error))
 
+async function containerIsRunning(container: string) {
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply content/src/entrypoints/run-local-database.ts:17]: why in the `catalyst-owner`? this entrypoint is not used afaik. I tried running it locally and works

File: content/src/entrypoints/run-maintenance.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  async main(program: Lifecycle.EntryPointParameters<MaintenanceComponents>): Promise<void> {
+    const { components, startComponents, stop } = program
+
+    await components.migrationManager.run()
+
+    await migrateContentFolderStructure(components)
+
+    await deleteUnreferencedFiles(components)
+
+    await startComponents()
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply content/src/entrypoints/run-maintenance.ts:23]: yes, moved but only before the delete. The migrate must be run before the start

File: content/src/migrations/ContentFolderMigrationManager.ts [hunk 6]
```diff
@@ -57,6 +60,7 @@ export async function migrateContentFolderStructure(components: ContentFolderMig
 
   try {
     await Promise.all(queued)
+    logs.info(`Migrated ${migratedCount} files`)
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply content/src/migrations/ContentFolderMigrationManager.ts:63]: not possible as there is no running server :(

File: content/src/entrypoints/run-maintenance.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import { createDatabaseComponent } from '../ports/postgres'
+import { MaintenanceComponents } from '../types'
+
+void Lifecycle.run({
+  async main(program: Lifecycle.EntryPointParameters<MaintenanceComponents>): Promise<void> {
+    const { components, startComponents, stop } = program
+
+    await components.migrationManager.run()
+
+    await migrateContentFolderStructure(components)
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply content/src/entrypoints/run-maintenance.ts:21]: can't add metrics as stated in other comment. But this was already fixed to log only when files are moved. (there was a bug reporting bad numbers)

File: content/src/logic/delete-unreferenced-files.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pedrotambo) [reply content/src/logic/delete-unreferenced-files.ts]: why to stderr?

File: content/src/entrypoints/run-maintenance.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    await startComponents()
+
+    await deleteUnreferencedFiles(components)
+
+    await stop()
+  },
+
+  async initComponents() {
+    const logs = createLogComponent()
+    const metrics = createTestMetricsComponent(metricsDeclaration)
```
Comment (agusaldasoro) [inline content/src/entrypoints/run-maintenance.ts:32]: do we need metrics in maintenance mode?

File: content/test/integration/logic/delete-unreferenced-files.spec.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    await server.startProgram()
+    const deployResult = await buildDeployData(['0,0', '0,1'], {
+      metadata: 'this is just some metadata'
+    })
+
+    await server.deploy(deployResult.deployData)
+
+    const contentsByHash: Map<ContentFileHash, Uint8Array> = await ServiceImpl.hashFiles(deployResult.deployData.files, deployResult.deployData.entityId)
+    const contentFileHashes = Array.from(contentsByHash.keys())
+    const unreferencedFileHash = 'a-hash'
```
Comment (agusaldasoro) [inline content/test/integration/logic/delete-unreferenced-files.spec.ts:48]: do you need creating this file for this test?

File: content/test/integration/logic/delete-unreferenced-files.spec.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    const components = server.components
+    makeNoopValidator(components)
+    await server.startProgram()
+    const deployResult = await buildDeployData(['0,0', '0,1'], {
+      metadata: 'this is just some metadata'
+    })
+
+    await server.deploy(deployResult.deployData)
+
+    const contentsByHash: Map<ContentFileHash, Uint8Array> = await ServiceImpl.hashFiles(deployResult.deployData.files, deployResult.deployData.entityId)
```
Comment (agusaldasoro) [inline content/test/integration/logic/delete-unreferenced-files.spec.ts:46]: if files is empty why are you hashing it?

File: content/test/integration/logic/delete-unreferenced-files.spec.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        metadata: 'this is just some metadata',
+        contentPaths: [getIntegrationResourcePathFor('some-binary-file.png')]
+      })
+
+      await server.deploy(deployResult.deployData)
+
+      const contentsByHash: Map<ContentFileHash, Uint8Array> = await ServiceImpl.hashFiles(deployResult.deployData.files, deployResult.deployData.entityId)
+      const contentFileHashes = Array.from(contentsByHash.keys())
+
+      const unreferencedFileHash = 'a-hash'
```
Comment (agusaldasoro) [inline content/test/integration/logic/delete-unreferenced-files.spec.ts:73]: same here, maybe creating this files doesn't add anything to this test

**Review by agusaldasoro — APPROVED** [review]
Great work!!

File: content/src/entrypoints/run-maintenance.ts (hunk 8 above)
Comment (pedrotambo) [reply content/src/entrypoints/run-maintenance.ts:32]: Nope.. but the db component need it for its creation :(. Should i create a fake one?

File: content/test/integration/logic/delete-unreferenced-files.spec.ts (hunk 9 above)
Comment (pedrotambo) [reply content/test/integration/logic/delete-unreferenced-files.spec.ts:48]: It is not fully needed... just to verify that it `does` delete unreferenced files but `not` entity ids

File: content/test/integration/logic/delete-unreferenced-files.spec.ts (hunk 10 above)
Comment (pedrotambo) [reply content/test/integration/logic/delete-unreferenced-files.spec.ts:46]: `Files` contains the content_files but also de entity files

File: content/test/integration/logic/delete-unreferenced-files.spec.ts (hunk 11 above)
Comment (pedrotambo) [reply content/test/integration/logic/delete-unreferenced-files.spec.ts:73]: Yeah.. i just wanted to cover the case where there are `referenced` and `not referenced` files, and it does delete only the `unreferenced` ones. Maybe i could remove these lines from these tests and add another one

Comment (pedrotambo) [reply content/test/integration/logic/delete-unreferenced-files.spec.ts]: Refactored.

**Pull request merged**